crates/rpc/rpc/src/eth/pubsub.rs

```diff
@@ -292,22 +289,13 @@ where
 {
     /// Returns a stream that yields all new RPC blocks.
     fn new_headers_stream(&self) -> impl Stream<Item = Header> {
-        BroadcastStream::new(self.chain_events.subscribe_to_canonical_state())
-            .map(|new_block| {
-                let new_chain = new_block.expect("new block subscription never ends; qed");
-                new_chain
-                    .committed()
-                    .map(|c| {
-                        c.blocks()
-                            .iter()
-                            .map(|(_, block)| {
-                                Header::from_primitive_with_hash(block.header.clone())
-                            })
-                            .collect::<Vec<_>>()
-                    })
-                    .unwrap_or_default()
-            })
-            .flat_map(futures::stream::iter)
+        self.chain_events.canonical_state_stream().flat_map(|new_chain| {
+            let headers = new_chain
+                .committed()
+                .map(|chain| chain.headers().collect::<Vec<_>>())
+                .unwrap_or_default();
+            futures::stream::iter(headers.into_iter().map(Header::from_primitive_with_hash))
```

Comment:
nit: is it possible to:

1. do the `Header::from_primitive_with_hash` in the `chain.headers()` portion
2. do `.unwrap_or(iter::empty())`

to avoid the `.collect() -> .into_iter()` part?